// @react-aria/visually-hidden/dist/module.js
import {mergeProps as $48FbJ$mergeProps} from "@react-aria/utils";
import $48FbJ$react, {useState as $48FbJ$useState, useMemo as $48FbJ$useMemo} from "react";
import {useFocus as $48FbJ$useFocus} from "@react-aria/interactions";

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $b05beb39a168f0a6$exports = {};

$parcel$export($b05beb39a168f0a6$exports, "useVisuallyHidden", () => $b05beb39a168f0a6$export$a966af930f325cab);
$parcel$export($b05beb39a168f0a6$exports, "VisuallyHidden", () => $b05beb39a168f0a6$export$439d29a4e110a164);



const $b05beb39a168f0a6$var$styles = {
    border: 0,
    clip: 'rect(0 0 0 0)',
    clipPath: 'inset(50%)',
    height: 1,
    margin: '0 -1px -1px 0',
    overflow: 'hidden',
    padding: 0,
    position: 'absolute',
    width: 1,
    whiteSpace: 'nowrap'
};
function $b05beb39a168f0a6$export$a966af930f325cab(props = {
}) {
    let { style: style , isFocusable: isFocusable  } = props;
    let [isFocused, setFocused] = $48FbJ$useState(false);
    let { focusProps: focusProps  } = $48FbJ$useFocus({
        isDisabled: !isFocusable,
        onFocusChange: setFocused
    });
    // If focused, don't hide the element.
    let combinedStyles = $48FbJ$useMemo(()=>{
        if (isFocused) return style;
        else if (style) return {
            ...$b05beb39a168f0a6$var$styles,
            ...style
        };
        else return $b05beb39a168f0a6$var$styles;
    }, [
        isFocused
    ]);
    return {
        visuallyHiddenProps: {
            ...focusProps,
            style: combinedStyles
        }
    };
}
function $b05beb39a168f0a6$export$439d29a4e110a164(props) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    let { children: children , elementType: Element = 'div' , isFocusable: isFocusable , style: style , ...otherProps } = props;
    let { visuallyHiddenProps: visuallyHiddenProps  } = $b05beb39a168f0a6$export$a966af930f325cab(props);
    return(/*#__PURE__*/ $48FbJ$react.createElement(Element, $48FbJ$mergeProps(otherProps, visuallyHiddenProps), children));
}




export {$b05beb39a168f0a6$export$a966af930f325cab as useVisuallyHidden, $b05beb39a168f0a6$export$439d29a4e110a164 as VisuallyHidden};
