// @react-aria/focus/dist/module.js
import $bmn4K$react, {useRef as $bmn4K$useRef, useContext as $bmn4K$useContext, useEffect as $bmn4K$useEffect, useState as $bmn4K$useState} from "react";
import {useLayoutEffect as $bmn4K$useLayoutEffect, runAfterTransition as $bmn4K$runAfterTransition, focusWithoutScrolling as $bmn4K$focusWithoutScrolling, mergeProps as $bmn4K$mergeProps, useSyncRef as $bmn4K$useSyncRef} from "@react-aria/utils";
import {getInteractionModality as $bmn4K$getInteractionModality, isFocusVisible as $bmn4K$isFocusVisible, useFocusVisibleListener as $bmn4K$useFocusVisibleListener, useFocus as $bmn4K$useFocus, useFocusWithin as $bmn4K$useFocusWithin, useKeyboard as $bmn4K$useKeyboard} from "@react-aria/interactions";
import $bmn4K$clsx from "clsx";

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $0238e4b796d1715f$exports = {};

$parcel$export($0238e4b796d1715f$exports, "FocusScope", () => $0238e4b796d1715f$export$20e40289641fbbb6);
$parcel$export($0238e4b796d1715f$exports, "useFocusManager", () => $0238e4b796d1715f$export$10c5169755ce7bd7);
$parcel$export($0238e4b796d1715f$exports, "getFocusableTreeWalker", () => $0238e4b796d1715f$export$2d6ec8fc375ceafa);
$parcel$export($0238e4b796d1715f$exports, "createFocusManager", () => $0238e4b796d1715f$export$c5251b9e124bf29);
var $f2c1256fdcfd2c09$exports = {};

$parcel$export($f2c1256fdcfd2c09$exports, "focusSafely", () => $f2c1256fdcfd2c09$export$80f3e147d781571c);


function $f2c1256fdcfd2c09$export$80f3e147d781571c(element) {
    // If the user is interacting with a virtual cursor, e.g. screen reader, then
    // wait until after any animated transitions that are currently occurring on
    // the page before shifting focus. This avoids issues with VoiceOver on iOS
    // causing the page to scroll when moving focus if the element is transitioning
    // from off the screen.
    if ($bmn4K$getInteractionModality() === 'virtual') {
        let lastFocusedElement = document.activeElement;
        $bmn4K$runAfterTransition(()=>{
            // If focus did not move and the element is still in the document, focus it.
            if (document.activeElement === lastFocusedElement && document.contains(element)) $bmn4K$focusWithoutScrolling(element);
        });
    } else $bmn4K$focusWithoutScrolling(element);
}


/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ function $c5775c2d6be42e6d$var$isStyleVisible(element) {
    if (!(element instanceof HTMLElement) && !(element instanceof SVGElement)) return false;
    let { display: display , visibility: visibility  } = element.style;
    let isVisible = display !== 'none' && visibility !== 'hidden' && visibility !== 'collapse';
    if (isVisible) {
        const { getComputedStyle: getComputedStyle  } = element.ownerDocument.defaultView;
        let { display: computedDisplay , visibility: computedVisibility  } = getComputedStyle(element);
        isVisible = computedDisplay !== 'none' && computedVisibility !== 'hidden' && computedVisibility !== 'collapse';
    }
    return isVisible;
}
function $c5775c2d6be42e6d$var$isAttributeVisible(element, childElement) {
    return !element.hasAttribute('hidden') && (element.nodeName === 'DETAILS' && childElement && childElement.nodeName !== 'SUMMARY' ? element.hasAttribute('open') : true);
}
function $c5775c2d6be42e6d$export$e989c0fffaa6b27a(element, childElement) {
    return element.nodeName !== '#comment' && $c5775c2d6be42e6d$var$isStyleVisible(element) && $c5775c2d6be42e6d$var$isAttributeVisible(element, childElement) && (!element.parentElement || $c5775c2d6be42e6d$export$e989c0fffaa6b27a(element.parentElement, element));
}




const $0238e4b796d1715f$var$FocusContext = /*#__PURE__*/ $bmn4K$react.createContext(null);
let $0238e4b796d1715f$var$activeScope = null;
let $0238e4b796d1715f$var$scopes = new Map();
function $0238e4b796d1715f$export$20e40289641fbbb6(props) {
    let { children: children , contain: contain , restoreFocus: restoreFocus , autoFocus: autoFocus  } = props;
    let startRef = $bmn4K$useRef();
    let endRef = $bmn4K$useRef();
    let scopeRef = $bmn4K$useRef([]);
    let ctx = $bmn4K$useContext($0238e4b796d1715f$var$FocusContext);
    let parentScope = ctx?.scopeRef;
    $bmn4K$useLayoutEffect(()=>{
        // Find all rendered nodes between the sentinels and add them to the scope.
        let node = startRef.current.nextSibling;
        let nodes = [];
        while(node && node !== endRef.current){
            nodes.push(node);
            node = node.nextSibling;
        }
        scopeRef.current = nodes;
    }, [
        children,
        parentScope
    ]);
    $bmn4K$useLayoutEffect(()=>{
        $0238e4b796d1715f$var$scopes.set(scopeRef, parentScope);
        return ()=>{
            // Restore the active scope on unmount if this scope or a descendant scope is active.
            // Parent effect cleanups run before children, so we need to check if the
            // parent scope actually still exists before restoring the active scope to it.
            if ((scopeRef === $0238e4b796d1715f$var$activeScope || $0238e4b796d1715f$var$isAncestorScope(scopeRef, $0238e4b796d1715f$var$activeScope)) && (!parentScope || $0238e4b796d1715f$var$scopes.has(parentScope))) $0238e4b796d1715f$var$activeScope = parentScope;
            $0238e4b796d1715f$var$scopes.delete(scopeRef);
        };
    }, [
        scopeRef,
        parentScope
    ]);
    $0238e4b796d1715f$var$useFocusContainment(scopeRef, contain);
    $0238e4b796d1715f$var$useRestoreFocus(scopeRef, restoreFocus, contain);
    $0238e4b796d1715f$var$useAutoFocus(scopeRef, autoFocus);
    let focusManager = $0238e4b796d1715f$var$createFocusManagerForScope(scopeRef);
    return(/*#__PURE__*/ $bmn4K$react.createElement($0238e4b796d1715f$var$FocusContext.Provider, {
        value: {
            scopeRef: scopeRef,
            focusManager: focusManager
        }
    }, /*#__PURE__*/ $bmn4K$react.createElement("span", {
        "data-focus-scope-start": true,
        hidden: true,
        ref: startRef
    }), children, /*#__PURE__*/ $bmn4K$react.createElement("span", {
        "data-focus-scope-end": true,
        hidden: true,
        ref: endRef
    })));
}
function $0238e4b796d1715f$export$10c5169755ce7bd7() {
    return $bmn4K$useContext($0238e4b796d1715f$var$FocusContext)?.focusManager;
}
function $0238e4b796d1715f$var$createFocusManagerForScope(scopeRef) {
    return {
        focusNext (opts = {
        }) {
            let scope = scopeRef.current;
            let { from: from , tabbable: tabbable , wrap: wrap  } = opts;
            let node = from || document.activeElement;
            let sentinel = scope[0].previousElementSibling;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
                tabbable: tabbable
            }, scope);
            walker.currentNode = $0238e4b796d1715f$var$isElementInScope(node, scope) ? node : sentinel;
            let nextNode = walker.nextNode();
            if (!nextNode && wrap) {
                walker.currentNode = sentinel;
                nextNode = walker.nextNode();
            }
            if (nextNode) $0238e4b796d1715f$var$focusElement(nextNode, true);
            return nextNode;
        },
        focusPrevious (opts = {
        }) {
            let scope = scopeRef.current;
            let { from: from , tabbable: tabbable , wrap: wrap  } = opts;
            let node = from || document.activeElement;
            let sentinel = scope[scope.length - 1].nextElementSibling;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
                tabbable: tabbable
            }, scope);
            walker.currentNode = $0238e4b796d1715f$var$isElementInScope(node, scope) ? node : sentinel;
            let previousNode = walker.previousNode();
            if (!previousNode && wrap) {
                walker.currentNode = sentinel;
                previousNode = walker.previousNode();
            }
            if (previousNode) $0238e4b796d1715f$var$focusElement(previousNode, true);
            return previousNode;
        },
        focusFirst (opts = {
        }) {
            let scope = scopeRef.current;
            let { tabbable: tabbable  } = opts;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
                tabbable: tabbable
            }, scope);
            walker.currentNode = scope[0].previousElementSibling;
            let nextNode = walker.nextNode();
            if (nextNode) $0238e4b796d1715f$var$focusElement(nextNode, true);
            return nextNode;
        },
        focusLast (opts = {
        }) {
            let scope = scopeRef.current;
            let { tabbable: tabbable  } = opts;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
                tabbable: tabbable
            }, scope);
            walker.currentNode = scope[scope.length - 1].nextElementSibling;
            let previousNode = walker.previousNode();
            if (previousNode) $0238e4b796d1715f$var$focusElement(previousNode, true);
            return previousNode;
        }
    };
}
const $0238e4b796d1715f$var$focusableElements = [
    'input:not([disabled]):not([type=hidden])',
    'select:not([disabled])',
    'textarea:not([disabled])',
    'button:not([disabled])',
    'a[href]',
    'area[href]',
    'summary',
    'iframe',
    'object',
    'embed',
    'audio[controls]',
    'video[controls]',
    '[contenteditable]'
];
const $0238e4b796d1715f$var$FOCUSABLE_ELEMENT_SELECTOR = $0238e4b796d1715f$var$focusableElements.join(':not([hidden]),') + ',[tabindex]:not([disabled]):not([hidden])';
$0238e4b796d1715f$var$focusableElements.push('[tabindex]:not([tabindex="-1"]):not([disabled])');
const $0238e4b796d1715f$var$TABBABLE_ELEMENT_SELECTOR = $0238e4b796d1715f$var$focusableElements.join(':not([hidden]):not([tabindex="-1"]),');
function $0238e4b796d1715f$var$getScopeRoot(scope) {
    return scope[0].parentElement;
}
function $0238e4b796d1715f$var$useFocusContainment(scopeRef, contain) {
    let focusedNode = $bmn4K$useRef();
    let raf = $bmn4K$useRef(null);
    $bmn4K$useLayoutEffect(()=>{
        let scope1 = scopeRef.current;
        if (!contain) return;
        // Handle the Tab key to contain focus within the scope
        let onKeyDown = (e)=>{
            if (e.key !== 'Tab' || e.altKey || e.ctrlKey || e.metaKey || scopeRef !== $0238e4b796d1715f$var$activeScope) return;
            let focusedElement = document.activeElement;
            let scope = scopeRef.current;
            if (!$0238e4b796d1715f$var$isElementInScope(focusedElement, scope)) return;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
                tabbable: true
            }, scope);
            walker.currentNode = focusedElement;
            let nextElement = e.shiftKey ? walker.previousNode() : walker.nextNode();
            if (!nextElement) {
                walker.currentNode = e.shiftKey ? scope[scope.length - 1].nextElementSibling : scope[0].previousElementSibling;
                nextElement = e.shiftKey ? walker.previousNode() : walker.nextNode();
            }
            e.preventDefault();
            if (nextElement) $0238e4b796d1715f$var$focusElement(nextElement, true);
        };
        let onFocus = (e)=>{
            // If focusing an element in a child scope of the currently active scope, the child becomes active.
            // Moving out of the active scope to an ancestor is not allowed.
            if (!$0238e4b796d1715f$var$activeScope || $0238e4b796d1715f$var$isAncestorScope($0238e4b796d1715f$var$activeScope, scopeRef)) {
                $0238e4b796d1715f$var$activeScope = scopeRef;
                focusedNode.current = e.target;
            } else if (scopeRef === $0238e4b796d1715f$var$activeScope && !$0238e4b796d1715f$var$isElementInChildScope(e.target, scopeRef)) {
                // If a focus event occurs outside the active scope (e.g. user tabs from browser location bar),
                // restore focus to the previously focused node or the first tabbable element in the active scope.
                if (focusedNode.current) focusedNode.current.focus();
                else if ($0238e4b796d1715f$var$activeScope) $0238e4b796d1715f$var$focusFirstInScope($0238e4b796d1715f$var$activeScope.current);
            } else if (scopeRef === $0238e4b796d1715f$var$activeScope) focusedNode.current = e.target;
        };
        let onBlur = (e)=>{
            // Firefox doesn't shift focus back to the Dialog properly without this
            raf.current = requestAnimationFrame(()=>{
                // Use document.activeElement instead of e.relatedTarget so we can tell if user clicked into iframe
                if (scopeRef === $0238e4b796d1715f$var$activeScope && !$0238e4b796d1715f$var$isElementInChildScope(document.activeElement, scopeRef)) {
                    $0238e4b796d1715f$var$activeScope = scopeRef;
                    focusedNode.current = e.target;
                    focusedNode.current.focus();
                }
            });
        };
        document.addEventListener('keydown', onKeyDown, false);
        document.addEventListener('focusin', onFocus, false);
        scope1.forEach((element)=>element.addEventListener('focusin', onFocus, false)
        );
        scope1.forEach((element)=>element.addEventListener('focusout', onBlur, false)
        );
        return ()=>{
            document.removeEventListener('keydown', onKeyDown, false);
            document.removeEventListener('focusin', onFocus, false);
            scope1.forEach((element)=>element.removeEventListener('focusin', onFocus, false)
            );
            scope1.forEach((element)=>element.removeEventListener('focusout', onBlur, false)
            );
        };
    }, [
        scopeRef,
        contain
    ]);
    // eslint-disable-next-line arrow-body-style
    $bmn4K$useEffect(()=>{
        return ()=>cancelAnimationFrame(raf.current)
        ;
    }, [
        raf
    ]);
}
function $0238e4b796d1715f$var$isElementInAnyScope(element) {
    for (let scope of $0238e4b796d1715f$var$scopes.keys()){
        if ($0238e4b796d1715f$var$isElementInScope(element, scope.current)) return true;
    }
    return false;
}
function $0238e4b796d1715f$var$isElementInScope(element, scope) {
    return scope.some((node)=>node.contains(element)
    );
}
function $0238e4b796d1715f$var$isElementInChildScope(element, scope) {
    // node.contains in isElementInScope covers child scopes that are also DOM children,
    // but does not cover child scopes in portals.
    for (let s of $0238e4b796d1715f$var$scopes.keys()){
        if ((s === scope || $0238e4b796d1715f$var$isAncestorScope(scope, s)) && $0238e4b796d1715f$var$isElementInScope(element, s.current)) return true;
    }
    return false;
}
function $0238e4b796d1715f$var$isAncestorScope(ancestor, scope) {
    let parent = $0238e4b796d1715f$var$scopes.get(scope);
    if (!parent) return false;
    if (parent === ancestor) return true;
    return $0238e4b796d1715f$var$isAncestorScope(ancestor, parent);
}
function $0238e4b796d1715f$var$focusElement(element, scroll = false) {
    if (element != null && !scroll) try {
        $f2c1256fdcfd2c09$export$80f3e147d781571c(element);
    } catch (err) {
    // ignore
    }
    else if (element != null) try {
        element.focus();
    } catch (err1) {
    // ignore
    }
}
function $0238e4b796d1715f$var$focusFirstInScope(scope) {
    let sentinel = scope[0].previousElementSibling;
    let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa($0238e4b796d1715f$var$getScopeRoot(scope), {
        tabbable: true
    }, scope);
    walker.currentNode = sentinel;
    $0238e4b796d1715f$var$focusElement(walker.nextNode());
}
function $0238e4b796d1715f$var$useAutoFocus(scopeRef, autoFocus) {
    const autoFocusRef = $bmn4K$react.useRef(autoFocus);
    $bmn4K$useEffect(()=>{
        if (autoFocusRef.current) {
            $0238e4b796d1715f$var$activeScope = scopeRef;
            if (!$0238e4b796d1715f$var$isElementInScope(document.activeElement, $0238e4b796d1715f$var$activeScope.current)) $0238e4b796d1715f$var$focusFirstInScope(scopeRef.current);
        }
        autoFocusRef.current = false;
    }, []);
}
function $0238e4b796d1715f$var$useRestoreFocus(scopeRef, restoreFocus, contain) {
    // create a ref during render instead of useLayoutEffect so the active element is saved before a child with autoFocus=true mounts.
    const nodeToRestoreRef = $bmn4K$useRef(typeof document !== 'undefined' ? document.activeElement : null);
    // useLayoutEffect instead of useEffect so the active element is saved synchronously instead of asynchronously.
    $bmn4K$useLayoutEffect(()=>{
        let nodeToRestore = nodeToRestoreRef.current;
        if (!restoreFocus) return;
        // Handle the Tab key so that tabbing out of the scope goes to the next element
        // after the node that had focus when the scope mounted. This is important when
        // using portals for overlays, so that focus goes to the expected element when
        // tabbing out of the overlay.
        let onKeyDown = (e)=>{
            if (e.key !== 'Tab' || e.altKey || e.ctrlKey || e.metaKey) return;
            let focusedElement = document.activeElement;
            if (!$0238e4b796d1715f$var$isElementInScope(focusedElement, scopeRef.current)) return;
            // Create a DOM tree walker that matches all tabbable elements
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa(document.body, {
                tabbable: true
            });
            // Find the next tabbable element after the currently focused element
            walker.currentNode = focusedElement;
            let nextElement = e.shiftKey ? walker.previousNode() : walker.nextNode();
            if (!document.body.contains(nodeToRestore) || nodeToRestore === document.body) nodeToRestore = null;
            // If there is no next element, or it is outside the current scope, move focus to the
            // next element after the node to restore to instead.
            if ((!nextElement || !$0238e4b796d1715f$var$isElementInScope(nextElement, scopeRef.current)) && nodeToRestore) {
                walker.currentNode = nodeToRestore;
                // Skip over elements within the scope, in case the scope immediately follows the node to restore.
                do nextElement = e.shiftKey ? walker.previousNode() : walker.nextNode();
                while ($0238e4b796d1715f$var$isElementInScope(nextElement, scopeRef.current))
                e.preventDefault();
                e.stopPropagation();
                if (nextElement) $0238e4b796d1715f$var$focusElement(nextElement, true);
                else // If there is no next element and the nodeToRestore isn't within a FocusScope (i.e. we are leaving the top level focus scope)
                // then move focus to the body.
                // Otherwise restore focus to the nodeToRestore (e.g menu within a popover -> tabbing to close the menu should move focus to menu trigger)
                if (!$0238e4b796d1715f$var$isElementInAnyScope(nodeToRestore)) focusedElement.blur();
                else $0238e4b796d1715f$var$focusElement(nodeToRestore, true);
            }
        };
        if (!contain) document.addEventListener('keydown', onKeyDown, true);
        return ()=>{
            if (!contain) document.removeEventListener('keydown', onKeyDown, true);
            if (restoreFocus && nodeToRestore && $0238e4b796d1715f$var$isElementInScope(document.activeElement, scopeRef.current)) requestAnimationFrame(()=>{
                if (document.body.contains(nodeToRestore)) $0238e4b796d1715f$var$focusElement(nodeToRestore);
            });
        };
    }, [
        scopeRef,
        restoreFocus,
        contain
    ]);
}
function $0238e4b796d1715f$export$2d6ec8fc375ceafa(root, opts, scope) {
    let selector = opts?.tabbable ? $0238e4b796d1715f$var$TABBABLE_ELEMENT_SELECTOR : $0238e4b796d1715f$var$FOCUSABLE_ELEMENT_SELECTOR;
    let walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, {
        acceptNode (node) {
            // Skip nodes inside the starting node.
            if (opts?.from?.contains(node)) return NodeFilter.FILTER_REJECT;
            if (node.matches(selector) && $c5775c2d6be42e6d$export$e989c0fffaa6b27a(node) && (!scope || $0238e4b796d1715f$var$isElementInScope(node, scope))) return NodeFilter.FILTER_ACCEPT;
            return NodeFilter.FILTER_SKIP;
        }
    });
    if (opts?.from) walker.currentNode = opts.from;
    return walker;
}
function $0238e4b796d1715f$export$c5251b9e124bf29(ref) {
    return {
        focusNext (opts = {
        }) {
            let root = ref.current;
            let { from: from , tabbable: tabbable , wrap: wrap  } = opts;
            let node = from || document.activeElement;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa(root, {
                tabbable: tabbable
            });
            if (root.contains(node)) walker.currentNode = node;
            let nextNode = walker.nextNode();
            if (!nextNode && wrap) {
                walker.currentNode = root;
                nextNode = walker.nextNode();
            }
            if (nextNode) $0238e4b796d1715f$var$focusElement(nextNode, true);
            return nextNode;
        },
        focusPrevious (opts = {
        }) {
            let root = ref.current;
            let { from: from , tabbable: tabbable , wrap: wrap  } = opts;
            let node = from || document.activeElement;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa(root, {
                tabbable: tabbable
            });
            if (root.contains(node)) walker.currentNode = node;
            else {
                let next = $0238e4b796d1715f$var$last(walker);
                if (next) $0238e4b796d1715f$var$focusElement(next, true);
                return next;
            }
            let previousNode = walker.previousNode();
            if (!previousNode && wrap) {
                walker.currentNode = root;
                previousNode = $0238e4b796d1715f$var$last(walker);
            }
            if (previousNode) $0238e4b796d1715f$var$focusElement(previousNode, true);
            return previousNode;
        },
        focusFirst (opts = {
        }) {
            let root = ref.current;
            let { tabbable: tabbable  } = opts;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa(root, {
                tabbable: tabbable
            });
            let nextNode = walker.nextNode();
            if (nextNode) $0238e4b796d1715f$var$focusElement(nextNode, true);
            return nextNode;
        },
        focusLast (opts = {
        }) {
            let root = ref.current;
            let { tabbable: tabbable  } = opts;
            let walker = $0238e4b796d1715f$export$2d6ec8fc375ceafa(root, {
                tabbable: tabbable
            });
            let next = $0238e4b796d1715f$var$last(walker);
            if (next) $0238e4b796d1715f$var$focusElement(next, true);
            return next;
        }
    };
}
function $0238e4b796d1715f$var$last(walker) {
    let next;
    let last;
    do {
        last = walker.lastChild();
        if (last) next = last;
    }while (last)
    return next;
}


var $db88f84346fe3322$exports = {};

$parcel$export($db88f84346fe3322$exports, "FocusRing", () => $db88f84346fe3322$export$1a38b4ad7f578e1d);



var $e974583017b16a4e$exports = {};

$parcel$export($e974583017b16a4e$exports, "useFocusRing", () => $e974583017b16a4e$export$4e328f61c538687f);



function $e974583017b16a4e$export$4e328f61c538687f(props = {
}) {
    let { autoFocus: autoFocus = false , isTextInput: isTextInput , within: within  } = props;
    let state = $bmn4K$useRef({
        isFocused: false,
        isFocusVisible: autoFocus || $bmn4K$isFocusVisible()
    }).current;
    let [isFocused1, setFocused] = $bmn4K$useState(false);
    let [isFocusVisibleState, setFocusVisible] = $bmn4K$useState(()=>state.isFocused && state.isFocusVisible
    );
    let updateState = ()=>setFocusVisible(state.isFocused && state.isFocusVisible)
    ;
    let onFocusChange = (isFocused)=>{
        state.isFocused = isFocused;
        setFocused(isFocused);
        updateState();
    };
    $bmn4K$useFocusVisibleListener((isFocusVisible)=>{
        state.isFocusVisible = isFocusVisible;
        updateState();
    }, [], {
        isTextInput: isTextInput
    });
    let { focusProps: focusProps  } = $bmn4K$useFocus({
        isDisabled: within,
        onFocusChange: onFocusChange
    });
    let { focusWithinProps: focusWithinProps  } = $bmn4K$useFocusWithin({
        isDisabled: !within,
        onFocusWithinChange: onFocusChange
    });
    return {
        isFocused: isFocused1,
        isFocusVisible: state.isFocused && isFocusVisibleState,
        focusProps: within ? focusWithinProps : focusProps
    };
}


function $db88f84346fe3322$export$1a38b4ad7f578e1d(props) {
    let { children: children , focusClass: focusClass , focusRingClass: focusRingClass  } = props;
    let { isFocused: isFocused , isFocusVisible: isFocusVisible , focusProps: focusProps  } = $e974583017b16a4e$export$4e328f61c538687f(props);
    let child = $bmn4K$react.Children.only(children);
    return(/*#__PURE__*/ $bmn4K$react.cloneElement(child, $bmn4K$mergeProps(child.props, {
        ...focusProps,
        className: $bmn4K$clsx({
            [focusClass || '']: isFocused,
            [focusRingClass || '']: isFocusVisible
        })
    })));
}


var $e82d8245213b3ebf$exports = {};

$parcel$export($e82d8245213b3ebf$exports, "FocusableProvider", () => $e82d8245213b3ebf$export$13f3202a3e5ddd5);
$parcel$export($e82d8245213b3ebf$exports, "useFocusable", () => $e82d8245213b3ebf$export$4c014de7c8940b4c);



let $e82d8245213b3ebf$var$FocusableContext = /*#__PURE__*/ $bmn4K$react.createContext(null);
function $e82d8245213b3ebf$var$useFocusableContext(ref) {
    let context = $bmn4K$useContext($e82d8245213b3ebf$var$FocusableContext) || {
    };
    $bmn4K$useSyncRef(context, ref);
    // eslint-disable-next-line
    let { ref: _ , ...otherProps } = context;
    return otherProps;
}
/**
 * Provides DOM props to the nearest focusable child.
 */ function $e82d8245213b3ebf$var$FocusableProvider(props, ref) {
    let { children: children , ...otherProps } = props;
    let context = {
        ...otherProps,
        ref: ref
    };
    return(/*#__PURE__*/ $bmn4K$react.createElement($e82d8245213b3ebf$var$FocusableContext.Provider, {
        value: context
    }, children));
}
let $e82d8245213b3ebf$export$13f3202a3e5ddd5 = /*#__PURE__*/ $bmn4K$react.forwardRef($e82d8245213b3ebf$var$FocusableProvider);
function $e82d8245213b3ebf$export$4c014de7c8940b4c(props, domRef) {
    let { focusProps: focusProps  } = $bmn4K$useFocus(props);
    let { keyboardProps: keyboardProps  } = $bmn4K$useKeyboard(props);
    let interactions = $bmn4K$mergeProps(focusProps, keyboardProps);
    let domProps = $e82d8245213b3ebf$var$useFocusableContext(domRef);
    let interactionProps = props.isDisabled ? {
    } : domProps;
    let autoFocusRef = $bmn4K$useRef(props.autoFocus);
    $bmn4K$useEffect(()=>{
        if (autoFocusRef.current && domRef.current) domRef.current.focus();
        autoFocusRef.current = false;
    }, []);
    return {
        focusableProps: $bmn4K$mergeProps({
            ...interactions,
            tabIndex: props.excludeFromTabOrder && !props.isDisabled ? -1 : undefined
        }, interactionProps)
    };
}






export {$0238e4b796d1715f$export$20e40289641fbbb6 as FocusScope, $0238e4b796d1715f$export$10c5169755ce7bd7 as useFocusManager, $0238e4b796d1715f$export$2d6ec8fc375ceafa as getFocusableTreeWalker, $0238e4b796d1715f$export$c5251b9e124bf29 as createFocusManager, $db88f84346fe3322$export$1a38b4ad7f578e1d as FocusRing, $e82d8245213b3ebf$export$13f3202a3e5ddd5 as FocusableProvider, $e82d8245213b3ebf$export$4c014de7c8940b4c as useFocusable, $e974583017b16a4e$export$4e328f61c538687f as useFocusRing, $f2c1256fdcfd2c09$export$80f3e147d781571c as focusSafely};
