// @react-aria/checkbox/dist/main.js
var $csc5b$react = require("react");
var $csc5b$reactariatoggle = require("@react-aria/toggle");
var $csc5b$reactariautils = require("@react-aria/utils");
var $csc5b$reactarialabel = require("@react-aria/label");
var $csc5b$reactstatelytoggle = require("@react-stately/toggle");

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $8e3a8695775dae64$exports = {};

$parcel$export($8e3a8695775dae64$exports, "useCheckbox", () => $8e3a8695775dae64$export$e375f10ce42261c5);


function $8e3a8695775dae64$export$e375f10ce42261c5(props, state, inputRef) {
    let { inputProps: inputProps  } = $csc5b$reactariatoggle.useToggle(props, state, inputRef);
    let { isSelected: isSelected  } = state;
    let { isIndeterminate: isIndeterminate  } = props;
    $csc5b$react.useEffect(()=>{
        // indeterminate is a property, but it can only be set via javascript
        // https://css-tricks.com/indeterminate-checkboxes/
        if (inputRef.current) inputRef.current.indeterminate = isIndeterminate;
    });
    return {
        inputProps: {
            ...inputProps,
            checked: isSelected,
            'aria-checked': isIndeterminate ? 'mixed' : isSelected
        }
    };
}


var $236317cbe6d4d104$exports = {};

$parcel$export($236317cbe6d4d104$exports, "useCheckboxGroup", () => $236317cbe6d4d104$export$49ff6f28c54f1cbe);
const $494dde44f66cf558$export$31440636951aa68c = new WeakMap();




function $236317cbe6d4d104$export$49ff6f28c54f1cbe(props, state) {
    let { isDisabled: isDisabled , name: name  } = props;
    let { labelProps: labelProps , fieldProps: fieldProps  } = $csc5b$reactarialabel.useLabel({
        ...props,
        // Checkbox group is not an HTML input element so it
        // shouldn't be labeled by a <label> element.
        labelElementType: 'span'
    });
    let domProps = $csc5b$reactariautils.filterDOMProps(props, {
        labelable: true
    });
    // Pass name prop from group to all items by attaching to the state.
    $494dde44f66cf558$export$31440636951aa68c.set(state, name);
    return {
        groupProps: $csc5b$reactariautils.mergeProps(domProps, {
            role: 'group',
            'aria-disabled': isDisabled || undefined,
            ...fieldProps
        }),
        labelProps: labelProps
    };
}


var $40670ca1adde6ca2$exports = {};

$parcel$export($40670ca1adde6ca2$exports, "useCheckboxGroupItem", () => $40670ca1adde6ca2$export$353b32fc6898d37d);



function $40670ca1adde6ca2$export$353b32fc6898d37d(props, state, inputRef) {
    const toggleState = $csc5b$reactstatelytoggle.useToggleState({
        isReadOnly: props.isReadOnly || state.isReadOnly,
        isSelected: state.isSelected(props.value),
        onChange (isSelected) {
            if (isSelected) state.addValue(props.value);
            else state.removeValue(props.value);
            if (props.onChange) props.onChange(isSelected);
        }
    });
    let { inputProps: inputProps  } = $8e3a8695775dae64$export$e375f10ce42261c5({
        ...props,
        isReadOnly: props.isReadOnly || state.isReadOnly,
        isDisabled: props.isDisabled || state.isDisabled,
        name: props.name || $494dde44f66cf558$export$31440636951aa68c.get(state)
    }, toggleState, inputRef);
    return {
        inputProps: inputProps
    };
}


$parcel$exportWildcard(module.exports, $8e3a8695775dae64$exports);
$parcel$exportWildcard(module.exports, $236317cbe6d4d104$exports);
$parcel$exportWildcard(module.exports, $40670ca1adde6ca2$exports);
