// @react-aria/checkbox/dist/module.js
import {useEffect as $lS2ep$useEffect} from "react";
import {useToggle as $lS2ep$useToggle} from "@react-aria/toggle";
import {filterDOMProps as $lS2ep$filterDOMProps, mergeProps as $lS2ep$mergeProps} from "@react-aria/utils";
import {useLabel as $lS2ep$useLabel} from "@react-aria/label";
import {useToggleState as $lS2ep$useToggleState} from "@react-stately/toggle";

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $6fc5544ca1b3ab2a$exports = {};

$parcel$export($6fc5544ca1b3ab2a$exports, "useCheckbox", () => $6fc5544ca1b3ab2a$export$e375f10ce42261c5);


function $6fc5544ca1b3ab2a$export$e375f10ce42261c5(props, state, inputRef) {
    let { inputProps: inputProps  } = $lS2ep$useToggle(props, state, inputRef);
    let { isSelected: isSelected  } = state;
    let { isIndeterminate: isIndeterminate  } = props;
    $lS2ep$useEffect(()=>{
        // indeterminate is a property, but it can only be set via javascript
        // https://css-tricks.com/indeterminate-checkboxes/
        if (inputRef.current) inputRef.current.indeterminate = isIndeterminate;
    });
    return {
        inputProps: {
            ...inputProps,
            checked: isSelected,
            'aria-checked': isIndeterminate ? 'mixed' : isSelected
        }
    };
}


var $9c576869c2451f7f$exports = {};

$parcel$export($9c576869c2451f7f$exports, "useCheckboxGroup", () => $9c576869c2451f7f$export$49ff6f28c54f1cbe);
const $46cd8c3ac90d2083$export$31440636951aa68c = new WeakMap();




function $9c576869c2451f7f$export$49ff6f28c54f1cbe(props, state) {
    let { isDisabled: isDisabled , name: name  } = props;
    let { labelProps: labelProps , fieldProps: fieldProps  } = $lS2ep$useLabel({
        ...props,
        // Checkbox group is not an HTML input element so it
        // shouldn't be labeled by a <label> element.
        labelElementType: 'span'
    });
    let domProps = $lS2ep$filterDOMProps(props, {
        labelable: true
    });
    // Pass name prop from group to all items by attaching to the state.
    $46cd8c3ac90d2083$export$31440636951aa68c.set(state, name);
    return {
        groupProps: $lS2ep$mergeProps(domProps, {
            role: 'group',
            'aria-disabled': isDisabled || undefined,
            ...fieldProps
        }),
        labelProps: labelProps
    };
}


var $1e517f900339691c$exports = {};

$parcel$export($1e517f900339691c$exports, "useCheckboxGroupItem", () => $1e517f900339691c$export$353b32fc6898d37d);



function $1e517f900339691c$export$353b32fc6898d37d(props, state, inputRef) {
    const toggleState = $lS2ep$useToggleState({
        isReadOnly: props.isReadOnly || state.isReadOnly,
        isSelected: state.isSelected(props.value),
        onChange (isSelected) {
            if (isSelected) state.addValue(props.value);
            else state.removeValue(props.value);
            if (props.onChange) props.onChange(isSelected);
        }
    });
    let { inputProps: inputProps  } = $6fc5544ca1b3ab2a$export$e375f10ce42261c5({
        ...props,
        isReadOnly: props.isReadOnly || state.isReadOnly,
        isDisabled: props.isDisabled || state.isDisabled,
        name: props.name || $46cd8c3ac90d2083$export$31440636951aa68c.get(state)
    }, toggleState, inputRef);
    return {
        inputProps: inputProps
    };
}




export {$6fc5544ca1b3ab2a$export$e375f10ce42261c5 as useCheckbox, $9c576869c2451f7f$export$49ff6f28c54f1cbe as useCheckboxGroup, $1e517f900339691c$export$353b32fc6898d37d as useCheckboxGroupItem};
