// @react-aria/utils/dist/main.js
var $gqa0x$react = require("react");
var $gqa0x$reactariassr = require("@react-aria/ssr");
var $gqa0x$clsx = require("clsx");
var $gqa0x$reactstatelyutils = require("@react-stately/utils");

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
var $dd96e9bc07c5b25f$exports = {};

$parcel$export($dd96e9bc07c5b25f$exports, "useId", () => $dd96e9bc07c5b25f$export$f680877a34711e37);
$parcel$export($dd96e9bc07c5b25f$exports, "mergeIds", () => $dd96e9bc07c5b25f$export$cd8c9cb68f842629);
$parcel$export($dd96e9bc07c5b25f$exports, "useSlotId", () => $dd96e9bc07c5b25f$export$b4cc09c592e8fdb8);

var $565383b8cde869dd$exports = {};

$parcel$export($565383b8cde869dd$exports, "useLayoutEffect", () => $565383b8cde869dd$export$e5c5a5f917a5871c);

const $565383b8cde869dd$export$e5c5a5f917a5871c = typeof window !== 'undefined' ? ($parcel$interopDefault($gqa0x$react)).useLayoutEffect : ()=>{
};




let $dd96e9bc07c5b25f$var$idsUpdaterMap = new Map();
function $dd96e9bc07c5b25f$export$f680877a34711e37(defaultId) {
    let isRendering = $gqa0x$react.useRef(true);
    isRendering.current = true;
    let [value, setValue] = $gqa0x$react.useState(defaultId);
    let nextId = $gqa0x$react.useRef(null);
    let res = $gqa0x$reactariassr.useSSRSafeId(value);
    // don't memo this, we want it new each render so that the Effects always run
    let updateValue = (val)=>{
        if (!isRendering.current) setValue(val);
        else nextId.current = val;
    };
    $dd96e9bc07c5b25f$var$idsUpdaterMap.set(res, updateValue);
    $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        isRendering.current = false;
    }, [
        updateValue
    ]);
    $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        let r = res;
        return ()=>{
            $dd96e9bc07c5b25f$var$idsUpdaterMap.delete(r);
        };
    }, [
        res
    ]);
    $gqa0x$react.useEffect(()=>{
        let newId = nextId.current;
        if (newId) {
            setValue(newId);
            nextId.current = null;
        }
    }, [
        setValue,
        updateValue
    ]);
    return res;
}
function $dd96e9bc07c5b25f$export$cd8c9cb68f842629(idA, idB) {
    if (idA === idB) return idA;
    let setIdA = $dd96e9bc07c5b25f$var$idsUpdaterMap.get(idA);
    if (setIdA) {
        setIdA(idB);
        return idB;
    }
    let setIdB = $dd96e9bc07c5b25f$var$idsUpdaterMap.get(idB);
    if (setIdB) {
        setIdB(idA);
        return idA;
    }
    return idB;
}
function $dd96e9bc07c5b25f$export$b4cc09c592e8fdb8(depArray = []) {
    let id = $dd96e9bc07c5b25f$export$f680877a34711e37();
    let [resolvedId, setResolvedId] = $789b9514fd01258b$export$14d238f342723f25(id);
    let updateId = $gqa0x$react.useCallback(()=>{
        setResolvedId(function*() {
            yield id;
            yield document.getElementById(id) ? id : null;
        });
    }, [
        id,
        setResolvedId
    ]);
    $565383b8cde869dd$export$e5c5a5f917a5871c(updateId, [
        id,
        updateId,
        ...depArray
    ]);
    return resolvedId;
}


var $5172289b7a1cda90$exports = {};

$parcel$export($5172289b7a1cda90$exports, "chain", () => $5172289b7a1cda90$export$e08e3b67e392101e);
function $5172289b7a1cda90$export$e08e3b67e392101e(...callbacks) {
    return (...args)=>{
        for (let callback of callbacks)if (typeof callback === 'function') callback(...args);
    };
}


var $ee1eca53264e4e21$exports = {};

$parcel$export($ee1eca53264e4e21$exports, "mergeProps", () => $ee1eca53264e4e21$export$9d1611c77c2fe928);



function $ee1eca53264e4e21$export$9d1611c77c2fe928(...args) {
    // Start with a base clone of the first argument. This is a lot faster than starting
    // with an empty object and adding properties as we go.
    let result = {
        ...args[0]
    };
    for(let i = 1; i < args.length; i++){
        let props = args[i];
        for(let key in props){
            let a = result[key];
            let b = props[key];
            // Chain events
            if (typeof a === 'function' && typeof b === 'function' && // This is a lot faster than a regex.
            key[0] === 'o' && key[1] === 'n' && key.charCodeAt(2) >= /* 'A' */ 65 && key.charCodeAt(2) <= /* 'Z' */ 90) result[key] = $5172289b7a1cda90$export$e08e3b67e392101e(a, b);
            else if ((key === 'className' || key === 'UNSAFE_className') && typeof a === 'string' && typeof b === 'string') result[key] = ($parcel$interopDefault($gqa0x$clsx))(a, b);
            else if (key === 'id' && a && b) result.id = $dd96e9bc07c5b25f$export$cd8c9cb68f842629(a, b);
            else result[key] = b !== undefined ? b : a;
        }
    }
    return result;
}


var $2f71b7bbfef4ef86$exports = {};

$parcel$export($2f71b7bbfef4ef86$exports, "filterDOMProps", () => $2f71b7bbfef4ef86$export$457c3d6518dd4c6f);
const $2f71b7bbfef4ef86$var$DOMPropNames = new Set([
    'id'
]);
const $2f71b7bbfef4ef86$var$labelablePropNames = new Set([
    'aria-label',
    'aria-labelledby',
    'aria-describedby',
    'aria-details'
]);
const $2f71b7bbfef4ef86$var$propRe = /^(data-.*)$/;
function $2f71b7bbfef4ef86$export$457c3d6518dd4c6f(props, opts = {
}) {
    let { labelable: labelable , propNames: propNames  } = opts;
    let filteredProps = {
    };
    for(const prop in props)if (Object.prototype.hasOwnProperty.call(props, prop) && ($2f71b7bbfef4ef86$var$DOMPropNames.has(prop) || labelable && $2f71b7bbfef4ef86$var$labelablePropNames.has(prop) || propNames?.has(prop) || $2f71b7bbfef4ef86$var$propRe.test(prop))) filteredProps[prop] = props[prop];
    return filteredProps;
}


var $448ba02dcac1c856$exports = {};

$parcel$export($448ba02dcac1c856$exports, "focusWithoutScrolling", () => $448ba02dcac1c856$export$de79e2c695e052f3);
function $448ba02dcac1c856$export$de79e2c695e052f3(element) {
    if ($448ba02dcac1c856$var$supportsPreventScroll()) element.focus({
        preventScroll: true
    });
    else {
        let scrollableElements = $448ba02dcac1c856$var$getScrollableElements(element);
        element.focus();
        $448ba02dcac1c856$var$restoreScrollPosition(scrollableElements);
    }
}
let $448ba02dcac1c856$var$supportsPreventScrollCached = null;
function $448ba02dcac1c856$var$supportsPreventScroll() {
    if ($448ba02dcac1c856$var$supportsPreventScrollCached == null) {
        $448ba02dcac1c856$var$supportsPreventScrollCached = false;
        try {
            var focusElem = document.createElement('div');
            focusElem.focus({
                get preventScroll () {
                    $448ba02dcac1c856$var$supportsPreventScrollCached = true;
                    return true;
                }
            });
        } catch (e) {
        // Ignore
        }
    }
    return $448ba02dcac1c856$var$supportsPreventScrollCached;
}
function $448ba02dcac1c856$var$getScrollableElements(element) {
    var parent = element.parentNode;
    var scrollableElements = [];
    var rootScrollingElement = document.scrollingElement || document.documentElement;
    while(parent instanceof HTMLElement && parent !== rootScrollingElement){
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) scrollableElements.push({
            element: parent,
            scrollTop: parent.scrollTop,
            scrollLeft: parent.scrollLeft
        });
        parent = parent.parentNode;
    }
    if (rootScrollingElement instanceof HTMLElement) scrollableElements.push({
        element: rootScrollingElement,
        scrollTop: rootScrollingElement.scrollTop,
        scrollLeft: rootScrollingElement.scrollLeft
    });
    return scrollableElements;
}
function $448ba02dcac1c856$var$restoreScrollPosition(scrollableElements) {
    for (let { element: element , scrollTop: scrollTop , scrollLeft: scrollLeft  } of scrollableElements){
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}


var $391ef47b89fbbdf8$exports = {};

$parcel$export($391ef47b89fbbdf8$exports, "getOffset", () => $391ef47b89fbbdf8$export$622cea445a1c5b7d);
function $391ef47b89fbbdf8$export$622cea445a1c5b7d(element, reverse, orientation = 'horizontal') {
    let rect = element.getBoundingClientRect();
    if (reverse) return orientation === 'horizontal' ? rect.right : rect.bottom;
    return orientation === 'horizontal' ? rect.left : rect.top;
}


var $c570a0a6fa1d588f$exports = {};

$parcel$export($c570a0a6fa1d588f$exports, "clamp", () => $gqa0x$reactstatelyutils.clamp);
$parcel$export($c570a0a6fa1d588f$exports, "snapValueToStep", () => $gqa0x$reactstatelyutils.snapValueToStep);



var $4da380280744869b$exports = {};

$parcel$export($4da380280744869b$exports, "runAfterTransition", () => $4da380280744869b$export$24490316f764c430);
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ // We store a global list of elements that are currently transitioning,
// mapped to a set of CSS properties that are transitioning for that element.
// This is necessary rather than a simple count of transitions because of browser
// bugs, e.g. Chrome sometimes fires both transitionend and transitioncancel rather
// than one or the other. So we need to track what's actually transitioning so that
// we can ignore these duplicate events.
let $4da380280744869b$var$transitionsByElement = new Map();
// A list of callbacks to call once there are no transitioning elements.
let $4da380280744869b$var$transitionCallbacks = new Set();
function $4da380280744869b$var$setupGlobalEvents() {
    if (typeof window === 'undefined') return;
    let onTransitionStart = (e)=>{
        // Add the transitioning property to the list for this element.
        let transitions = $4da380280744869b$var$transitionsByElement.get(e.target);
        if (!transitions) {
            transitions = new Set();
            $4da380280744869b$var$transitionsByElement.set(e.target, transitions);
            // The transitioncancel event must be registered on the element itself, rather than as a global
            // event. This enables us to handle when the node is deleted from the document while it is transitioning.
            // In that case, the cancel event would have nowhere to bubble to so we need to handle it directly.
            e.target.addEventListener('transitioncancel', onTransitionEnd);
        }
        transitions.add(e.propertyName);
    };
    let onTransitionEnd = (e)=>{
        // Remove property from list of transitioning properties.
        let properties = $4da380280744869b$var$transitionsByElement.get(e.target);
        if (!properties) return;
        properties.delete(e.propertyName);
        // If empty, remove transitioncancel event, and remove the element from the list of transitioning elements.
        if (properties.size === 0) {
            e.target.removeEventListener('transitioncancel', onTransitionEnd);
            $4da380280744869b$var$transitionsByElement.delete(e.target);
        }
        // If no transitioning elements, call all of the queued callbacks.
        if ($4da380280744869b$var$transitionsByElement.size === 0) {
            for (let cb of $4da380280744869b$var$transitionCallbacks)cb();
            $4da380280744869b$var$transitionCallbacks.clear();
        }
    };
    document.body.addEventListener('transitionrun', onTransitionStart);
    document.body.addEventListener('transitionend', onTransitionEnd);
}
if (typeof document !== 'undefined') {
    if (document.readyState !== 'loading') $4da380280744869b$var$setupGlobalEvents();
    else document.addEventListener('DOMContentLoaded', $4da380280744869b$var$setupGlobalEvents);
}
function $4da380280744869b$export$24490316f764c430(fn) {
    // Wait one frame to see if an animation starts, e.g. a transition on mount.
    requestAnimationFrame(()=>{
        // If no transitions are running, call the function immediately.
        // Otherwise, add it to a list of callbacks to run at the end of the animation.
        if ($4da380280744869b$var$transitionsByElement.size === 0) fn();
        else $4da380280744869b$var$transitionCallbacks.add(fn);
    });
}


var $ca85cc095ee7bdf9$exports = {};

$parcel$export($ca85cc095ee7bdf9$exports, "useDrag1D", () => $ca85cc095ee7bdf9$export$7bbed75feba39706);


// Keep track of elements that we are currently handling dragging for via useDrag1D.
// If there's an ancestor and a descendant both using useDrag1D(), and the user starts
// dragging the descendant, we don't want useDrag1D events to fire for the ancestor.
const $ca85cc095ee7bdf9$var$draggingElements = [];
function $ca85cc095ee7bdf9$export$7bbed75feba39706(props) {
    console.warn('useDrag1D is deprecated, please use `useMove` instead https://react-spectrum.adobe.com/react-aria/useMove.html');
    let { containerRef: containerRef , reverse: reverse , orientation: orientation , onHover: onHover , onDrag: onDrag , onPositionChange: onPositionChange , onIncrement: onIncrement , onDecrement: onDecrement , onIncrementToMax: onIncrementToMax , onDecrementToMin: onDecrementToMin , onCollapseToggle: onCollapseToggle  } = props;
    let getPosition = (e)=>orientation === 'horizontal' ? e.clientX : e.clientY
    ;
    let getNextOffset = (e)=>{
        let containerOffset = $391ef47b89fbbdf8$export$622cea445a1c5b7d(containerRef.current, reverse, orientation);
        let mouseOffset = getPosition(e);
        let nextOffset = reverse ? containerOffset - mouseOffset : mouseOffset - containerOffset;
        return nextOffset;
    };
    let dragging = $gqa0x$react.useRef(false);
    let prevPosition = $gqa0x$react.useRef(0);
    // Keep track of the current handlers in a ref so that the events can access them.
    let handlers = $gqa0x$react.useRef({
        onPositionChange: onPositionChange,
        onDrag: onDrag
    });
    handlers.current.onDrag = onDrag;
    handlers.current.onPositionChange = onPositionChange;
    let onMouseDragged = (e)=>{
        e.preventDefault();
        let nextOffset = getNextOffset(e);
        if (!dragging.current) {
            dragging.current = true;
            if (handlers.current.onDrag) handlers.current.onDrag(true);
            if (handlers.current.onPositionChange) handlers.current.onPositionChange(nextOffset);
        }
        if (prevPosition.current === nextOffset) return;
        prevPosition.current = nextOffset;
        if (onPositionChange) onPositionChange(nextOffset);
    };
    let onMouseUp = (e)=>{
        const target = e.target;
        dragging.current = false;
        let nextOffset = getNextOffset(e);
        if (handlers.current.onDrag) handlers.current.onDrag(false);
        if (handlers.current.onPositionChange) handlers.current.onPositionChange(nextOffset);
        $ca85cc095ee7bdf9$var$draggingElements.splice($ca85cc095ee7bdf9$var$draggingElements.indexOf(target), 1);
        window.removeEventListener('mouseup', onMouseUp, false);
        window.removeEventListener('mousemove', onMouseDragged, false);
    };
    let onMouseDown = (e)=>{
        const target = e.currentTarget;
        // If we're already handling dragging on a descendant with useDrag1D, then
        // we don't want to handle the drag motion on this target as well.
        if ($ca85cc095ee7bdf9$var$draggingElements.some((elt)=>target.contains(elt)
        )) return;
        $ca85cc095ee7bdf9$var$draggingElements.push(target);
        window.addEventListener('mousemove', onMouseDragged, false);
        window.addEventListener('mouseup', onMouseUp, false);
    };
    let onMouseEnter = ()=>{
        if (onHover) onHover(true);
    };
    let onMouseOut = ()=>{
        if (onHover) onHover(false);
    };
    let onKeyDown = (e)=>{
        switch(e.key){
            case 'Left':
            case 'ArrowLeft':
                if (orientation === 'horizontal') {
                    e.preventDefault();
                    if (onDecrement && !reverse) onDecrement();
                    else if (onIncrement && reverse) onIncrement();
                }
                break;
            case 'Up':
            case 'ArrowUp':
                if (orientation === 'vertical') {
                    e.preventDefault();
                    if (onDecrement && !reverse) onDecrement();
                    else if (onIncrement && reverse) onIncrement();
                }
                break;
            case 'Right':
            case 'ArrowRight':
                if (orientation === 'horizontal') {
                    e.preventDefault();
                    if (onIncrement && !reverse) onIncrement();
                    else if (onDecrement && reverse) onDecrement();
                }
                break;
            case 'Down':
            case 'ArrowDown':
                if (orientation === 'vertical') {
                    e.preventDefault();
                    if (onIncrement && !reverse) onIncrement();
                    else if (onDecrement && reverse) onDecrement();
                }
                break;
            case 'Home':
                e.preventDefault();
                if (onDecrementToMin) onDecrementToMin();
                break;
            case 'End':
                e.preventDefault();
                if (onIncrementToMax) onIncrementToMax();
                break;
            case 'Enter':
                e.preventDefault();
                if (onCollapseToggle) onCollapseToggle();
                break;
        }
    };
    return {
        onMouseDown: onMouseDown,
        onMouseEnter: onMouseEnter,
        onMouseOut: onMouseOut,
        onKeyDown: onKeyDown
    };
}


var $efa617098dfd7e72$exports = {};

$parcel$export($efa617098dfd7e72$exports, "useGlobalListeners", () => $efa617098dfd7e72$export$4eaf04e54aa8eed6);

function $efa617098dfd7e72$export$4eaf04e54aa8eed6() {
    let globalListeners = $gqa0x$react.useRef(new Map());
    let addGlobalListener = $gqa0x$react.useCallback((eventTarget, type, listener, options)=>{
        // Make sure we remove the listener after it is called with the `once` option.
        let fn = options?.once ? (...args)=>{
            globalListeners.current.delete(listener);
            listener(...args);
        } : listener;
        globalListeners.current.set(listener, {
            type: type,
            eventTarget: eventTarget,
            fn: fn,
            options: options
        });
        eventTarget.addEventListener(type, listener, options);
    }, []);
    let removeGlobalListener = $gqa0x$react.useCallback((eventTarget, type, listener, options)=>{
        let fn = globalListeners.current.get(listener)?.fn || listener;
        eventTarget.removeEventListener(type, fn, options);
        globalListeners.current.delete(listener);
    }, []);
    let removeAllGlobalListeners = $gqa0x$react.useCallback(()=>{
        globalListeners.current.forEach((value, key)=>{
            removeGlobalListener(value.eventTarget, value.type, key, value.options);
        });
    }, [
        removeGlobalListener
    ]);
    // eslint-disable-next-line arrow-body-style
    $gqa0x$react.useEffect(()=>{
        return removeAllGlobalListeners;
    }, [
        removeAllGlobalListeners
    ]);
    return {
        addGlobalListener: addGlobalListener,
        removeGlobalListener: removeGlobalListener,
        removeAllGlobalListeners: removeAllGlobalListeners
    };
}


var $e0460f7e24b931b0$exports = {};

$parcel$export($e0460f7e24b931b0$exports, "useLabels", () => $e0460f7e24b931b0$export$d6875122194c7b44);

function $e0460f7e24b931b0$export$d6875122194c7b44(props, defaultLabel) {
    let { id: id , 'aria-label': label , 'aria-labelledby': labelledBy  } = props;
    // If there is both an aria-label and aria-labelledby,
    // combine them by pointing to the element itself.
    id = $dd96e9bc07c5b25f$export$f680877a34711e37(id);
    if (labelledBy && label) {
        let ids = new Set([
            ...labelledBy.trim().split(/\s+/),
            id
        ]);
        labelledBy = [
            ...ids
        ].join(' ');
    } else if (labelledBy) labelledBy = labelledBy.trim().split(/\s+/).join(' ');
    // If no labels are provided, use the default
    if (!label && !labelledBy && defaultLabel) label = defaultLabel;
    return {
        id: id,
        'aria-label': label,
        'aria-labelledby': labelledBy
    };
}


var $658832d06264448f$exports = {};

$parcel$export($658832d06264448f$exports, "useObjectRef", () => $658832d06264448f$export$4338b53315abf666);


function $658832d06264448f$export$4338b53315abf666(forwardedRef) {
    const objRef = $gqa0x$react.useRef();
    /**
   * We're using `useLayoutEffect` here instead of `useEffect` because we want
   * to make sure that the `ref` value is up to date before other places in the
   * the execution cycle try to read it.
   */ $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        if (!forwardedRef) return;
        if (typeof forwardedRef === 'function') forwardedRef(objRef.current);
        else forwardedRef.current = objRef.current;
    }, [
        forwardedRef
    ]);
    return objRef;
}


var $5f37a52ce85585dd$exports = {};

$parcel$export($5f37a52ce85585dd$exports, "useUpdateEffect", () => $5f37a52ce85585dd$export$496315a1608d9602);

function $5f37a52ce85585dd$export$496315a1608d9602(effect, dependencies) {
    const isInitialMount = $gqa0x$react.useRef(true);
    $gqa0x$react.useEffect(()=>{
        if (isInitialMount.current) isInitialMount.current = false;
        else effect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, dependencies);
}



var $3599a41724a37362$exports = {};

$parcel$export($3599a41724a37362$exports, "useResizeObserver", () => $3599a41724a37362$export$683480f191c0e3ea);

function $3599a41724a37362$var$hasResizeObserver() {
    return typeof window.ResizeObserver !== 'undefined';
}
function $3599a41724a37362$export$683480f191c0e3ea(options) {
    const { ref: ref , onResize: onResize  } = options;
    $gqa0x$react.useEffect(()=>{
        let element = ref?.current;
        if (!element) return;
        if (!$3599a41724a37362$var$hasResizeObserver()) {
            window.addEventListener('resize', onResize, false);
            return ()=>{
                window.removeEventListener('resize', onResize, false);
            };
        } else {
            const resizeObserverInstance = new window.ResizeObserver((entries)=>{
                if (!entries.length) return;
                onResize();
            });
            resizeObserverInstance.observe(element);
            return ()=>{
                if (element) resizeObserverInstance.unobserve(element);
            };
        }
    }, [
        onResize,
        ref
    ]);
}


var $c5073b0bf61d031a$exports = {};

$parcel$export($c5073b0bf61d031a$exports, "useSyncRef", () => $c5073b0bf61d031a$export$4debdb1a3f0fa79e);

function $c5073b0bf61d031a$export$4debdb1a3f0fa79e(context, ref) {
    $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        if (context && context.ref && ref) {
            context.ref.current = ref.current;
            return ()=>{
                context.ref.current = null;
            };
        }
    }, [
        context,
        ref
    ]);
}


var $56cba3ac4ebbec36$exports = {};

$parcel$export($56cba3ac4ebbec36$exports, "getScrollParent", () => $56cba3ac4ebbec36$export$cfa2225e87938781);
function $56cba3ac4ebbec36$export$cfa2225e87938781(node) {
    while(node && !$56cba3ac4ebbec36$var$isScrollable(node))node = node.parentElement;
    return node || document.scrollingElement || document.documentElement;
}
function $56cba3ac4ebbec36$var$isScrollable(node) {
    let style = window.getComputedStyle(node);
    return /(auto|scroll)/.test(style.overflow + style.overflowX + style.overflowY);
}


var $eda1f9b13ac03ccc$exports = {};

$parcel$export($eda1f9b13ac03ccc$exports, "useViewportSize", () => $eda1f9b13ac03ccc$export$d699905dd57c73ca);

// @ts-ignore
let $eda1f9b13ac03ccc$var$visualViewport = typeof window !== 'undefined' && window.visualViewport;
function $eda1f9b13ac03ccc$export$d699905dd57c73ca() {
    let [size1, setSize] = $gqa0x$react.useState(()=>$eda1f9b13ac03ccc$var$getViewportSize()
    );
    $gqa0x$react.useEffect(()=>{
        // Use visualViewport api to track available height even on iOS virtual keyboard opening
        let onResize = ()=>{
            setSize((size)=>{
                let newSize = $eda1f9b13ac03ccc$var$getViewportSize();
                if (newSize.width === size.width && newSize.height === size.height) return size;
                return newSize;
            });
        };
        if (!$eda1f9b13ac03ccc$var$visualViewport) window.addEventListener('resize', onResize);
        else $eda1f9b13ac03ccc$var$visualViewport.addEventListener('resize', onResize);
        return ()=>{
            if (!$eda1f9b13ac03ccc$var$visualViewport) window.removeEventListener('resize', onResize);
            else $eda1f9b13ac03ccc$var$visualViewport.removeEventListener('resize', onResize);
        };
    }, []);
    return size1;
}
function $eda1f9b13ac03ccc$var$getViewportSize() {
    return {
        width: $eda1f9b13ac03ccc$var$visualViewport?.width || window.innerWidth,
        height: $eda1f9b13ac03ccc$var$visualViewport?.height || window.innerHeight
    };
}


var $fc64a6f8ffc8f0f9$exports = {};

$parcel$export($fc64a6f8ffc8f0f9$exports, "useDescription", () => $fc64a6f8ffc8f0f9$export$f8aeda7b10753fa1);


let $fc64a6f8ffc8f0f9$var$descriptionId = 0;
const $fc64a6f8ffc8f0f9$var$descriptionNodes = new Map();
function $fc64a6f8ffc8f0f9$export$f8aeda7b10753fa1(description) {
    let [id1, setId] = $gqa0x$react.useState(null);
    $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        if (!description) return;
        let desc = $fc64a6f8ffc8f0f9$var$descriptionNodes.get(description);
        if (!desc) {
            let id = `react-aria-description-${$fc64a6f8ffc8f0f9$var$descriptionId++}`;
            setId(id);
            let node = document.createElement('div');
            node.id = id;
            node.style.display = 'none';
            node.textContent = description;
            document.body.appendChild(node);
            desc = {
                refCount: 0,
                element: node
            };
            $fc64a6f8ffc8f0f9$var$descriptionNodes.set(description, desc);
        } else setId(desc.element.id);
        desc.refCount++;
        return ()=>{
            if (--desc.refCount === 0) {
                desc.element.remove();
                $fc64a6f8ffc8f0f9$var$descriptionNodes.delete(description);
            }
        };
    }, [
        description
    ]);
    return {
        'aria-describedby': description ? id1 : undefined
    };
}


var $8dce0acd60edf5c8$exports = {};

$parcel$export($8dce0acd60edf5c8$exports, "isMac", () => $8dce0acd60edf5c8$export$9ac100e40613ea10);
$parcel$export($8dce0acd60edf5c8$exports, "isIPhone", () => $8dce0acd60edf5c8$export$186c6964ca17d99);
$parcel$export($8dce0acd60edf5c8$exports, "isIPad", () => $8dce0acd60edf5c8$export$7bef049ce92e4224);
$parcel$export($8dce0acd60edf5c8$exports, "isIOS", () => $8dce0acd60edf5c8$export$fedb369cb70207f1);
$parcel$export($8dce0acd60edf5c8$exports, "isAppleDevice", () => $8dce0acd60edf5c8$export$e1865c3bedcd822b);
$parcel$export($8dce0acd60edf5c8$exports, "isWebKit", () => $8dce0acd60edf5c8$export$78551043582a6a98);
$parcel$export($8dce0acd60edf5c8$exports, "isChrome", () => $8dce0acd60edf5c8$export$6446a186d09e379e);
$parcel$export($8dce0acd60edf5c8$exports, "isAndroid", () => $8dce0acd60edf5c8$export$a11b0059900ceec8);
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */ function $8dce0acd60edf5c8$var$testUserAgent(re) {
    return typeof window !== 'undefined' && window.navigator != null ? re.test(window.navigator.userAgent) : false;
}
function $8dce0acd60edf5c8$var$testPlatform(re) {
    return typeof window !== 'undefined' && window.navigator != null ? re.test(window.navigator.platform) : false;
}
function $8dce0acd60edf5c8$export$9ac100e40613ea10() {
    return $8dce0acd60edf5c8$var$testPlatform(/^Mac/);
}
function $8dce0acd60edf5c8$export$186c6964ca17d99() {
    return $8dce0acd60edf5c8$var$testPlatform(/^iPhone/);
}
function $8dce0acd60edf5c8$export$7bef049ce92e4224() {
    return $8dce0acd60edf5c8$var$testPlatform(/^iPad/) || $8dce0acd60edf5c8$export$9ac100e40613ea10() && navigator.maxTouchPoints > 1;
}
function $8dce0acd60edf5c8$export$fedb369cb70207f1() {
    return $8dce0acd60edf5c8$export$186c6964ca17d99() || $8dce0acd60edf5c8$export$7bef049ce92e4224();
}
function $8dce0acd60edf5c8$export$e1865c3bedcd822b() {
    return $8dce0acd60edf5c8$export$9ac100e40613ea10() || $8dce0acd60edf5c8$export$fedb369cb70207f1();
}
function $8dce0acd60edf5c8$export$78551043582a6a98() {
    return $8dce0acd60edf5c8$var$testUserAgent(/AppleWebKit/) && !$8dce0acd60edf5c8$export$6446a186d09e379e();
}
function $8dce0acd60edf5c8$export$6446a186d09e379e() {
    return $8dce0acd60edf5c8$var$testUserAgent(/Chrome/);
}
function $8dce0acd60edf5c8$export$a11b0059900ceec8() {
    return $8dce0acd60edf5c8$var$testUserAgent(/Android/);
}


var $bc4cfb54fe7e6088$exports = {};

$parcel$export($bc4cfb54fe7e6088$exports, "useEvent", () => $bc4cfb54fe7e6088$export$90fc3a17d93f704c);

function $bc4cfb54fe7e6088$export$90fc3a17d93f704c(ref, event, handler1, options) {
    let handlerRef = $gqa0x$react.useRef(handler1);
    handlerRef.current = handler1;
    let isDisabled = handler1 == null;
    $gqa0x$react.useEffect(()=>{
        if (isDisabled) return;
        let element = ref.current;
        let handler = (e)=>handlerRef.current.call(this, e)
        ;
        element.addEventListener(event, handler, options);
        return ()=>{
            element.removeEventListener(event, handler, options);
        };
    }, [
        ref,
        event,
        options,
        isDisabled
    ]);
}


var $789b9514fd01258b$exports = {};

$parcel$export($789b9514fd01258b$exports, "useValueEffect", () => $789b9514fd01258b$export$14d238f342723f25);


function $789b9514fd01258b$export$14d238f342723f25(defaultValue) {
    let [value, setValue] = $gqa0x$react.useState(defaultValue);
    let valueRef = $gqa0x$react.useRef(value);
    let effect = $gqa0x$react.useRef(null);
    valueRef.current = value;
    // Store the function in a ref so we can always access the current version
    // which has the proper `value` in scope.
    let nextRef = $gqa0x$react.useRef(null);
    nextRef.current = ()=>{
        // Run the generator to the next yield.
        let newValue = effect.current.next();
        // If the generator is done, reset the effect.
        if (newValue.done) {
            effect.current = null;
            return;
        }
        // If the value is the same as the current value,
        // then continue to the next yield. Otherwise,
        // set the value in state and wait for the next layout effect.
        if (value === newValue.value) nextRef.current();
        else setValue(newValue.value);
    };
    $565383b8cde869dd$export$e5c5a5f917a5871c(()=>{
        // If there is an effect currently running, continue to the next yield.
        if (effect.current) nextRef.current();
    });
    let queue = $gqa0x$react.useCallback((fn)=>{
        effect.current = fn(valueRef.current);
        nextRef.current();
    }, [
        effect,
        nextRef
    ]);
    return [
        value,
        queue
    ];
}


var $c5c390447fdbf212$exports = {};

$parcel$export($c5c390447fdbf212$exports, "scrollIntoView", () => $c5c390447fdbf212$export$53a0910f038337bd);
function $c5c390447fdbf212$export$53a0910f038337bd(scrollView, element) {
    let offsetX = $c5c390447fdbf212$var$relativeOffset(scrollView, element, 'left');
    let offsetY = $c5c390447fdbf212$var$relativeOffset(scrollView, element, 'top');
    let width = element.offsetWidth;
    let height = element.offsetHeight;
    let x = scrollView.scrollLeft;
    let y = scrollView.scrollTop;
    let maxX = x + scrollView.offsetWidth;
    let maxY = y + scrollView.offsetHeight;
    if (offsetX <= x) x = offsetX;
    else if (offsetX + width > maxX) x += offsetX + width - maxX;
    if (offsetY <= y) y = offsetY;
    else if (offsetY + height > maxY) y += offsetY + height - maxY;
    scrollView.scrollLeft = x;
    scrollView.scrollTop = y;
}
/**
 * Computes the offset left or top from child to ancestor by accumulating
 * offsetLeft or offsetTop through intervening offsetParents.
 */ function $c5c390447fdbf212$var$relativeOffset(ancestor, child, axis) {
    const prop = axis === 'left' ? 'offsetLeft' : 'offsetTop';
    let sum = 0;
    while(child.offsetParent){
        sum += child[prop];
        if (child.offsetParent === ancestor) break;
        else if (child.offsetParent.contains(ancestor)) {
            // If the ancestor is not `position:relative`, then we stop at
            // _its_ offset parent, and we subtract off _its_ offset, so that
            // we end up with the proper offset from child to ancestor.
            sum -= ancestor[prop];
            break;
        }
        child = child.offsetParent;
    }
    return sum;
}


$parcel$exportWildcard(module.exports, $dd96e9bc07c5b25f$exports);
$parcel$exportWildcard(module.exports, $5172289b7a1cda90$exports);
$parcel$exportWildcard(module.exports, $ee1eca53264e4e21$exports);
$parcel$exportWildcard(module.exports, $2f71b7bbfef4ef86$exports);
$parcel$exportWildcard(module.exports, $448ba02dcac1c856$exports);
$parcel$exportWildcard(module.exports, $391ef47b89fbbdf8$exports);
$parcel$exportWildcard(module.exports, $c570a0a6fa1d588f$exports);
$parcel$exportWildcard(module.exports, $4da380280744869b$exports);
$parcel$exportWildcard(module.exports, $ca85cc095ee7bdf9$exports);
$parcel$exportWildcard(module.exports, $efa617098dfd7e72$exports);
$parcel$exportWildcard(module.exports, $e0460f7e24b931b0$exports);
$parcel$exportWildcard(module.exports, $658832d06264448f$exports);
$parcel$exportWildcard(module.exports, $5f37a52ce85585dd$exports);
$parcel$exportWildcard(module.exports, $565383b8cde869dd$exports);
$parcel$exportWildcard(module.exports, $3599a41724a37362$exports);
$parcel$exportWildcard(module.exports, $c5073b0bf61d031a$exports);
$parcel$exportWildcard(module.exports, $56cba3ac4ebbec36$exports);
$parcel$exportWildcard(module.exports, $eda1f9b13ac03ccc$exports);
$parcel$exportWildcard(module.exports, $fc64a6f8ffc8f0f9$exports);
$parcel$exportWildcard(module.exports, $8dce0acd60edf5c8$exports);
$parcel$exportWildcard(module.exports, $bc4cfb54fe7e6088$exports);
$parcel$exportWildcard(module.exports, $789b9514fd01258b$exports);
$parcel$exportWildcard(module.exports, $c5c390447fdbf212$exports);
