// @react-aria/label/dist/module.js
import {useSlotId as $5Odon$useSlotId, mergeProps as $5Odon$mergeProps, useId as $5Odon$useId, useLabels as $5Odon$useLabels} from "@react-aria/utils";

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $c796961545123cf9$exports = {};

$parcel$export($c796961545123cf9$exports, "useField", () => $c796961545123cf9$export$294aa081a6c6f55d);
var $a25d9b7703051f71$exports = {};

$parcel$export($a25d9b7703051f71$exports, "useLabel", () => $a25d9b7703051f71$export$8467354a121f1b9f);

function $a25d9b7703051f71$export$8467354a121f1b9f(props) {
    let { id: id , label: label , 'aria-labelledby': ariaLabelledby , 'aria-label': ariaLabel , labelElementType: labelElementType = 'label'  } = props;
    id = $5Odon$useId(id);
    let labelId = $5Odon$useId();
    let labelProps = {
    };
    if (label) {
        ariaLabelledby = ariaLabelledby ? `${ariaLabelledby} ${labelId}` : labelId;
        labelProps = {
            id: labelId,
            htmlFor: labelElementType === 'label' ? id : undefined
        };
    } else if (!ariaLabelledby && !ariaLabel) console.warn('If you do not provide a visible label, you must specify an aria-label or aria-labelledby attribute for accessibility');
    let fieldProps = $5Odon$useLabels({
        id: id,
        'aria-label': ariaLabel,
        'aria-labelledby': ariaLabelledby
    });
    return {
        labelProps: labelProps,
        fieldProps: fieldProps
    };
}



function $c796961545123cf9$export$294aa081a6c6f55d(props) {
    let { description: description , errorMessage: errorMessage , validationState: validationState  } = props;
    let { labelProps: labelProps , fieldProps: fieldProps  } = $a25d9b7703051f71$export$8467354a121f1b9f(props);
    let descriptionId = $5Odon$useSlotId([
        Boolean(description),
        Boolean(errorMessage),
        validationState
    ]);
    let errorMessageId = $5Odon$useSlotId([
        Boolean(description),
        Boolean(errorMessage),
        validationState
    ]);
    fieldProps = $5Odon$mergeProps(fieldProps, {
        'aria-describedby': [
            descriptionId,
            // Use aria-describedby for error message because aria-errormessage is unsupported using VoiceOver or NVDA. See https://github.com/adobe/react-spectrum/issues/1346#issuecomment-740136268
            errorMessageId,
            props['aria-describedby']
        ].filter(Boolean).join(' ') || undefined
    });
    return {
        labelProps: labelProps,
        fieldProps: fieldProps,
        descriptionProps: {
            id: descriptionId
        },
        errorMessageProps: {
            id: errorMessageId
        }
    };
}





export {$c796961545123cf9$export$294aa081a6c6f55d as useField, $a25d9b7703051f71$export$8467354a121f1b9f as useLabel};
